File: problem-solving/_lc-0647-medium-palindromic-substrings/1-solution.py
"""
https://leetcode.com/problems/palindromic-substrings/

Complexity
----------
Time: O(n^2)
Space: O(1)
"""

def countPalindrome(s, left, right):
    count = 0
    while left >= 0 and right < len(s) and s[left] == s[right]:
        left -= 1
        right += 1
        count += 1
    return count

class Solution:
    def countSubstrings(self, s: str) -> int:
        count = len(s)
        mid = 0
        while mid < len(s):
            count += countPalindrome(s, mid - 1, mid + 1)
            count += countPalindrome(s, mid, mid + 1)
            mid += 1
        return count

File: problem-solving/lc-5-medium-longest-palindromic-substring/1-solution.go
/*
https://leetcode.com/problems/longest-palindromic-substring/

Complexity
----------
Time: O(n^2)
Space: O(1)
*/

func isPalindrome(s string, l, r, start, end int) (int, int) {
    for l >= 0 && r < len(s) && s[l] == s[r] {
        if end - start < r - l {
            start, end = l, r
        }
        l -= 1
        r += 1
    }
    return start, end
}

func longestPalindrome(s string) string {
    start, end := 0, 0
    n := len(s)
    for i := 0; i < n; i++ {
        // Odd len palindrome - having center char at i
        start, end = isPalindrome(s, i - 1, i + 1, start, end)

        // Even len palindrome - don't have center char
        start, end = isPalindrome(s, i, i + 1, start, end)
    }
    return s[start:end + 1]
}
